util/stream_dataset.py

```diff
@@ -514,28 +514,54 @@ def interpolate_into(self, source_key, source_dataset, parameter):
         if source_key != self.stream_key:
             log.debug('<%s> interpolate_into: %s source: %s param: %r',
                       self.request_id, self.stream_key, source_key, parameter)
+            only_same_deployment = self.interpolate_only_from_same_deployment(source_key, parameter)
             new_name = '-'.join((source_key.stream.name, parameter.name))
             for deployment, ds in self.datasets.iteritems():
                 if new_name in ds:
                     continue
                 try:
-                    ds[new_name] = source_dataset.get_interpolated(ds.time.values, parameter)
+                    ds[new_name] = source_dataset.get_interpolated(ds.time.values, parameter,
+                                                                   deployment if only_same_deployment else None)
                 except StreamEngineException as e:
                     log.error(e.message)
 
+    def interpolate_only_from_same_deployment(self, source_key, parameter):
+        """
+        Determine whether we should interpolate data only from the same deployment
+        as the target dataset and not allow interpolation from across multiple
+        source datasets deployments.
+        :param source_key: RefDes + Stream for dataset to interpolate from
+        :param parameter: Parameter defining the data to be interpolated
+        :return: Boolean
+        """
+        # If the data level of the parameter in the source dataset is 0 (raw data)
+        # and the source and target datasets are from the same instrument,
+        # assume that data from the source can only be used from the same deployment
+        # as the target dataset that it is getting interpolated into.
+        # In contrast, a finished data product (data level greater than 0) and can be
+        # used interpolated into an different instrument's dataset regardless of
+        # deployment. Interpolating pressure form a CTD would be an example of the latter.
```

Comment:
The above assumptions sound reasonable to me, but I'm wondering if there are cases where the raw data file really represents L1 data and if that could cause a false negative here. I think there are some instruments or controllers that do a degree of pre-processing.

Reply:
Yes, I agree. However, to further prevent false negatives, we prevent interpolation across deployments only when the source data set and the target dataset are from the same instrument. The assumption here is that if a single instrument generates multiple data streams, those streams can be used by one another within the same deployment because those streams are the same physical instrument. If they are from different deployments, they may be from different physical instruments having a common reference designator and in that case a stream from one physical instrument probably should not depend on a stream from a different deployment, especially since that same supporting stream should also be available within the same deployment as the primary dataset.